casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -94,25 +81,40 @@ object ProtoUtil {
     * requires a list to be returned. When we reach the goalFunc,
     * we return an empty list.
     */
-  def getCreatorJustificationAsListUntilGoalInMemory[F[_]: Monad](
+  def getCreatorJustificationAsListUntilGoalInMemory[F[_]: Sync](
       blockDag: BlockDagRepresentation[F],
       blockHash: BlockHash,
       goalFunc: BlockHash => Boolean = _ => false
-  ): F[List[BlockHash]] =
-    (for {
-      block <- OptionT(blockDag.lookup(blockHash))
-      creatorJustificationHash <- OptionT.fromOption(
-                                   block.justifications
-                                     .find(_.validator == block.sender)
-                                     .map(_.latestBlockHash)
-                                 )
-      creatorJustification <- OptionT(blockDag.lookup(creatorJustificationHash))
-      creatorJustificationAsList = if (goalFunc(creatorJustification.blockHash)) {
-        List.empty[BlockHash]
-      } else {
-        List(creatorJustification.blockHash)
-      }
-    } yield creatorJustificationAsList).fold(List.empty[BlockHash])(identity)
+  ): F[List[BlockHash]] = {
+    import coop.rchain.blockstorage.syntax._
+    val getJ = (h: BlockHash) =>
+      blockDag
+        .lookupUnsafe(h)
+        .map(_.justifications.map { case Justification(m, v) => CliqueOracle.Justification(m, v) })
+        .map(_.toSet)
+    val toJ = (h: BlockHash) =>
+      blockDag
+        .lookupUnsafe(h)
+        .map(m => CliqueOracle.Justification(m.blockHash, m.sender))
```

Comment:
If `getCasperJustificationsUnsafe` and `toCasperJustificationUnsafe` will stay as BlockDagRepresentation extensions, they can be used here.

Reply:
I agree that this belongs to BlockMetadata syntax as you've pointed in https://github.com/rchain/rchain/pull/3387#discussion_r641605484